src/tree/block_scanner.rs

```diff
@@ -27,47 +34,77 @@ where
     pub async fn new(
         middleware: Arc<M>,
         window_size: u64,
-        current_block: u64,
+        start_block: u64,
         filter: Filter,
-    ) -> Result<Self, M::Error> {
+    ) -> WorldTreeResult<Self> {
         let chain_id = middleware.get_chainid().await?.as_u64();
         Ok(Self {
             middleware,
-            next_block: Mutex::new(current_block),
+            start_block,
             window_size,
             filter,
             chain_id,
         })
     }
 
-    /// Retrieves events matching the specified address and topics from the last synced block to the latest block, stepping by `window_size`.
-    /// Note that the logs are unsorted and should be handled accordingly.
-    pub async fn next(&self) -> Result<(usize, Vec<Log>), M::Error> {
-        let mut next_block = self.next_block.lock().await;
-        let latest_block = self.middleware.get_block_number().await?.as_u64();
-
-        let to_block = (*next_block + self.window_size).min(latest_block);
-
-        let num_blocks =
-            if let Some(num_blocks) = to_block.checked_sub(*next_block) {
-                num_blocks as usize
-            } else {
-                return Ok((0, vec![]));
+    pub fn block_stream(
+        &self,
+    ) -> impl Stream<Item: Future<Output = WorldTreeResult<Vec<Log>>> + Send> + '_
+    {
+        stream::unfold(self.start_block, move |mut next_block| async move {
+            // This is executed before the item is yielded
+            let to_block = loop {
+                match self.middleware.get_block_number().await {
+                    Ok(latest) if latest.as_u64() < next_block => {
+                        tokio::time::sleep(Duration::from_secs(
+                            BLOCK_SCANNER_SLEEP_TIME,
+                        ))
+                        .await;
+                        continue;
+                    }
+                    Ok(latest) => {
+                        break (next_block + self.window_size)
+                            .min(latest.as_u64());
+                    }
+                    Err(_) => {
+                        tokio::time::sleep(Duration::from_secs(
+                            BLOCK_SCANNER_SLEEP_TIME,
+                        ))
+                        .await;
+                        continue;
+                    }
+                }
             };
 
-        let filter = self
-            .filter
-            .clone()
-            .from_block(BlockNumber::Number((*next_block).into()))
-            .to_block(BlockNumber::Number(to_block.into()));
+            let filter = self
+                .filter
+                .clone()
```

Comment:
Consider simplifying to one match arm.

```suggestion
                Ok(latest) => {  
                        if latest.as_u64() < next_block {  
                            tokio::time::sleep(Duration::from_secs(  
                                BLOCK_SCANNER_SLEEP_TIME,  
                            ))  
                            .await;  
                            continue;  
                        } else{  
                            break (next_block + self.window_size)  
                                .min(latest.as_u64());  
                        }  
                        
                    }  
```

Reply:
Not a fan of `if` inside match arms eh?

Reply:
Either works, a single match arm is a bit more succinct here. 

If you feel strongly feel free to leave it!